File: vi/codes/javascript/chapter_array_and_linkedlist/my_list.js
/**
 * File: my_list.js
 * Created Time: 2022-12-12
 * Author: Justin (xiefahit@gmail.com)
 */

/* Lớp danh sách (list) */
class MyList {
    #arr = new Array(); // Mảng (lưu trữ các phần tử của danh sách)
    #capacity = 10; // Dung lượng của danh sách
    #size = 0; // Độ dài danh sách (số phần tử hiện tại)
    #extendRatio = 2; // Hệ số nhân dùng để mở rộng dung lượng danh sách mỗi lần

    /* Hàm khởi tạo */
    constructor() {
        this.#arr = new Array(this.#capacity);
    }

    /* Lấy độ dài danh sách (số phần tử hiện tại) */
    size() {
        return this.#size;
    }

    /* Lấy dung lượng của danh sách */
    capacity() {
        return this.#capacity;
    }

    /* Truy cập phần tử */
    get(index) {
        // Nếu chỉ số vượt quá phạm vi, ném ra ngoại lệ như bên dưới
        if (index < 0 || index >= this.#size) throw new Error('Index out of bounds');
        return this.#arr[index];
    }

    /* Cập nhật phần tử */
    set(index, num) {
        if (index < 0 || index >= this.#size) throw new Error('Index out of bounds');
        this.#arr[index] = num;
    }

    /* Thêm phần tử vào cuối danh sách */
    add(num) {
        // Nếu độ dài bằng dung lượng, cần mở rộng
        if (this.#size === this.#capacity) {
            this.extendCapacity();
        }
        // Thêm phần tử mới vào cuối danh sách
        this.#arr[this.#size] = num;
        this.#size++;
    }

    /* Chèn phần tử vào giữa danh sách */
    insert(index, num) {
        if (index < 0 || index >= this.#size) throw new Error('Index out of bounds');
        // Khi số phần tử vượt quá dung lượng, kích hoạt cơ chế mở rộng
        if (this.#size === this.#capacity) {
            this.extendCapacity();
        }
        // Dịch chuyển tất cả phần tử sau vị trí index tiến lên một vị trí
        for (let j = this.#size - 1; j >= index; j--) {
            this.#arr[j + 1] = this.#arr[j];
        }
        // Cập nhật số phần tử
        this.#arr[index] = num;
        this.#size++;
    }

    /* Xóa phần tử */
    remove(index) {
        if (index < 0 || index >= this.#size) throw new Error('Index out of bounds');
        let num = this.#arr[index];
        // Dịch chuyển tất cả phần tử sau vị trí index lùi lại một vị trí
        for (let j = index; j < this.#size - 1; j++) {
            this.#arr[j] = this.#arr[j + 1];
        }
        // Cập nhật số phần tử
        this.#size--;
        // Trả về phần tử đã bị xóa
        return num;
    }

    /* Mở rộng dung lượng danh sách */
    extendCapacity() {
        // Tạo một mảng mới có độ dài gấp extendRatio lần mảng gốc rồi sao chép mảng gốc sang mảng mới
        this.#arr = this.#arr.concat(
            new Array(this.capacity() * (this.#extendRatio - 1))
        );
        // Cập nhật dung lượng
        this.#capacity = this.#arr.length;
    }

    /* Chuyển danh sách thành mảng */
    toArray() {
        let size = this.size();
        // Chỉ chuyển đổi phạm vi hợp lệ trong độ dài size
        const arr = new Array(size);
        for (let i = 0; i < size; i++) {
            arr[i] = this.get(i);
        }
        return arr;
    }
}

/* Chương trình chính */
/* Khởi tạo danh sách */
const nums = new MyList();
/* Thêm phần tử vào cuối danh sách */
nums.add(1);
nums.add(3);
nums.add(2);
nums.add(5);
nums.add(4);
console.log(
    `List nums = ${nums.toArray()}, capacity = ${nums.capacity()}, length = ${nums.size()}`
);

/* Chèn phần tử vào giữa danh sách */
nums.insert(3, 6);
console.log(`Insert number 6 at index 3, get nums = ${nums.toArray()}`);

/* Xóa phần tử */
nums.remove(3);
console.log(`Delete element at index 3, get nums = ${nums.toArray()}`);

/* Truy cập phần tử */
const num = nums.get(1);
console.log(`Access element at index 1, get num = ${num}`);

/* Cập nhật phần tử */
nums.set(1, 0);
console.log(`Update element at index 1 to 0, get nums = ${nums.toArray()}`);

/* Kiểm tra cơ chế mở rộng dung lượng */
for (let i = 0; i < 10; i++) {
    // Khi i = 5, độ dài danh sách sẽ vượt quá dung lượng, kích hoạt cơ chế mở rộng
    nums.add(i);
}
console.log(
    `After expansion, list nums = ${nums.toArray()}, capacity = ${nums.capacity()}, length = ${nums.size()}`
);

File: vi/codes/ruby/chapter_array_and_linkedlist/my_list.rb
=begin
File: my_list.rb
Created Time: 2024-03-18
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### Lớp List ###
class MyList
  attr_reader :size       # Lấy độ dài list (số lượng phần tử hiện tại)
  attr_reader :capacity   # Lấy dung lượng list

  ### Hàm khởi tạo ###
  def initialize
    @capacity = 10
    @size = 0
    @extend_ratio = 2
    @arr = Array.new(capacity)
  end

  ### Truy cập phần tử ###
  def get(index)
    # Nếu chỉ số vượt quá giới hạn, ném ra ngoại lệ, như bên dưới
    raise IndexError, "Index out of bounds" if index < 0 || index >= size
    @arr[index]
  end

  ### Truy cập phần tử ###
  def set(index, num)
    raise IndexError, "Index out of bounds" if index < 0 || index >= size
    @arr[index] = num
  end

  ### Thêm phần tử vào cuối ###
  def add(num)
    # Khi số lượng phần tử vượt quá dung lượng, kích hoạt cơ chế mở rộng
    extend_capacity if size == capacity
    @arr[size] = num

    # Cập nhật số lượng phần tử
    @size += 1
  end

  ### Chèn phần tử vào giữa ###
  def insert(index, num)
    raise IndexError, "Index out of bounds" if index < 0 || index >= size

    # Khi số lượng phần tử vượt quá dung lượng, kích hoạt cơ chế mở rộng
    extend_capacity if size == capacity

    # Dịch tất cả phần tử sau vị trí index tiến lên một vị trí
    for j in (size - 1).downto(index)
      @arr[j + 1] = @arr[j]
    end
    @arr[index] = num

    # Cập nhật số lượng phần tử
    @size += 1
  end

  ### Xóa phần tử ###
  def remove(index)
    raise IndexError, "Index out of bounds" if index < 0 || index >= size
    num = @arr[index]

    # Dịch tất cả phần tử sau index tiến lên một vị trí
    for j in index...size
      @arr[j] = @arr[j + 1]
    end

    # Cập nhật số lượng phần tử
    @size -= 1

    # Trả về phần tử đã bị xóa
    num
  end

  ### Mở rộng dung lượng list ###
  def extend_capacity
    # Tạo mảng mới có độ dài gấp extend_ratio lần mảng gốc, sao chép mảng gốc sang mảng mới
    arr = @arr.dup + Array.new(capacity * (@extend_ratio - 1))
    # Thêm phần tử vào cuối
    @capacity = arr.length
  end

  ### Chuyển list thành mảng ###
  def to_array
    sz = size
    # Các phần tử vào hàng đợi
    arr = Array.new(sz)
    for i in 0...sz
      arr[i] = get(i)
    end
    arr
  end
end

### Driver Code ###
if __FILE__ == $0
  # Khởi tạo list
  nums = MyList.new

  # Duyệt trực tiếp các phần tử của list
  nums.add(1)
  nums.add(3)
  nums.add(2)
  nums.add(5)
  nums.add(4)
  puts "List nums = #{nums.to_array}, capacity = #{nums.capacity}, length = #{nums.size}"

  # Sắp xếp list
  nums.insert(3, 6)
  puts "Insert number 6 at index 3, get nums = #{nums.to_array}"

  # Xóa phần tử
  nums.remove(3)
  puts "Delete element at index 3, get nums = #{nums.to_array}"

  # Cập nhật phần tử
  num = nums.get(1)
  puts "Access element at index 1, get num = #{num}"

  # Thêm phần tử vào cuối
  nums.set(1, 0)
  puts "Update element at index 1 to 0, get nums = #{nums.to_array}"

  # Kiểm tra cơ chế mở rộng dung lượng
  for i in 0...10
    # Tại i = 5, độ dài list sẽ vượt quá dung lượng list, kích hoạt cơ chế mở rộng
    nums.add(i)
  end
  puts "After expansion, list nums = #{nums.to_array}, capacity = #{nums.capacity}, length = #{nums.size}"
end
</content>
</invoke>
